File: src-tauri/src/db/podcasts.rs
use super::*;
use serde::{Deserialize, Serialize};

/// One speaker in a podcast, and the voice that reads them.
///
/// The cast is written once from the script the model produced, then belongs to
/// the user: picking a voice, renaming a host, nudging rate or pitch. That is
/// why it is stored rather than derived — deriving it again on every open would
/// throw away every choice the moment anything upstream changed.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PodcastHost {
    /// The speaker name as it appears in the turns. This is the JOIN KEY
    /// between a line and a voice, which is why the script's own normalization
    /// matters so much: "Ada" in turn 1 and "Ada Lovelace" in turn 7 would be
    /// two hosts, and one of them would have no voice.
    pub name: String,
    /// A neural voice id from the live catalog (`tts.list_neural_voices`).
    /// Empty = the sidecar's product default, the same meaning it has in the
    /// Settings picker.
    #[serde(default)]
    pub voice: String,
    /// Edge prosody strings, e.g. "+22%" and "-2Hz". Empty = the service
    /// default for that voice. Per HOST, because two hosts reading at the
    /// identical rate is most of what makes a two-voice script still sound
    /// like one narrator.
    #[serde(default)]
    pub rate: String,
    #[serde(default)]
    pub pitch: String,
}

/// One line of the episode.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PodcastTurn {
    pub speaker: String,
    pub line: String,
}

/// A podcast script as data: what was written, who says it, and how it sounds.
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Podcast {
    /// The SCRIPT page's file id — the row's identity.
    pub file_id: String,
    pub title: String,
    pub turns: Vec<PodcastTurn>,
    pub cast: Vec<PodcastHost>,
    /// The rendered episode, when one has been made. An ordinary room file.
    pub audio_file_id: Option<String>,
    pub created_at: String,
}

/// Decoration a model wraps a speaker label in: `**Ada:**`, `[Ada]:`, `> Ada:`.
const DECOR: &[char] = &['*', '_', '[', ']', '#', '>', '"', '\'', '“', '”', ' ', '\t'];

/// Drop a leading `Speaker:` label from a line.
///
/// THE BUG THIS EXISTS FOR. Models write dialogue the way dialogue looks in
/// their training data — `{"speaker": "Ada", "line": "Ada: welcome in"}` — so
/// the name ends up stated twice: once as the field, once inside the text. The
/// field is the one everything joins on, and the copy inside the text is what
/// gets handed to the voice service. The result is a host who reads their own
/// name aloud before every line ("Ada: welcome in"), a page that prints the
/// name in the margin AND again in the sentence, and a transcript that reads
/// `Ada: Ada: welcome in`. One strip at the boundary fixes all three, because
/// all three read this one string.
///
/// WHY IT ONLY EVER STRIPS A NAME IT ALREADY KNOWS. The prefix must match this
/// turn's own speaker exactly (ignoring case and any decoration), or that
/// speaker's first word so `Ada Lovelace` still matches `Ada:`. A rule like
/// "cut at the first colon" would eat the start of every ordinary sentence
/// that happens to contain one — *"Here's the thing: it works"* — which is a
/// far worse failure than the one being fixed, and a silent one.
///
/// A line that is ONLY a label strips to empty; callers decide what that
/// means (ingest drops the turn, the read path keeps the original rather than
/// handing empty text to a speech service).
pub fn strip_speaker_label(line: &str, speaker: &str) -> String {
    let mut text = line.trim_start();
    // Bounded, not `while`: `Ada: Ada: hi` is real (a re-generated script fed
    // its own previous output), but an unbounded loop on adversarial input is
    // not worth the two lines it saves.
    for _ in 0..2 {
        match label_len(text, speaker) {
            Some(n) => text = text[n..].trim_start_matches(DECOR),
            None => break,
        }
    }
    text.to_string()
}

/// Byte length of a leading `Name:` run, or `None` if the line doesn't open
/// with this speaker's own name.
fn label_len(text: &str, speaker: &str) -> Option<usize> {
    /// A speaker label is short. Past this, a colon belongs to prose.
    const MAX_LABEL: usize = 48;
    let name = speaker.trim();
    if name.is_empty() {
        return None;
    }
    let colon = text
        .char_indices()
        .take_while(|(i, _)| *i <= MAX_LABEL)
        .find(|(_, c)| *c == ':')?
        .0;
    let prefix = text[..colon].trim_matches(DECOR);
    if prefix.is_empty() {
        return None;
    }
    let first_word = name.split_whitespace().next().unwrap_or(name);
    (prefix.eq_ignore_ascii_case(name) || prefix.eq_ignore_ascii_case(first_word))
        .then_some(colon + 1)
}

/// Strip every turn's own speaker label, in place.
///
/// A line that is nothing BUT a label keeps its original text: it is degenerate
/// either way, and empty text is the one thing the speech service cannot be
/// handed.
pub fn strip_speaker_labels(turns: &mut [PodcastTurn]) {
    for t in turns.iter_mut() {
        let cleaned = strip_speaker_label(&t.line, &t.speaker);
        if !cleaned.is_empty() {
            t.line = cleaned;
        }
    }
}

fn podcast_row(r: &rusqlite::Row<'_>) -> rusqlite::Result<Podcast> {
    let turns: String = r.get(2)?;
    let cast: String = r.get(3)?;
    // A row whose JSON will not parse is treated as an EMPTY script rather
    // than as a failed read. The page is still in the library and still
    // opens; what is lost is the ability to record it, and the panel says
    // so. Failing the whole query would take the working half down too.
    let mut turns: Vec<PodcastTurn> = serde_json::from_str(&turns).unwrap_or_default();
    // On the way OUT as well as in, so scripts written before the label strip
    // existed are fixed where they are read instead of needing a migration —
    // and so a re-record of an old episode stops speaking the names.
    strip_speaker_labels(&mut turns);
    Ok(Podcast {
        file_id: r.get(0)?,
        title: r.get(1)?,
        turns,
        cast: serde_json::from_str(&cast).unwrap_or_default(),
        audio_file_id: r.get(4)?,
        created_at: r.get(5)?,
    })
}

/// The podcast belonging to a script page, if it has one.
///
/// `None` covers two real cases the caller must not conflate with an error: a
/// file that is not a podcast at all, and a podcast PAGE generated before this
/// table existed (or by the old HTML-authoring path), which has no turns to
/// record.
pub fn get_podcast(conn: &Connection, file_id: &str) -> Option<Podcast> {
    conn.query_row(
        "SELECT file_id, title, turns, cast_json, audio_file_id, created_at
         FROM podcasts WHERE file_id = ?1",
        [file_id],
        podcast_row,
    )
    .ok()
}

/// Store (or replace) a script's structure. Called by the studio's `after_save`
/// hook with the model's own JSON already parsed.
pub fn save_podcast(
    conn: &Connection,
    file_id: &str,
    title: &str,
    turns: &[PodcastTurn],
    cast: &[PodcastHost],
) -> Result<(), String> {
    let turns = serde_json::to_string(turns).map_err(|e| e.to_string())?;
    let cast = serde_json::to_string(cast).map_err(|e| e.to_string())?;
    execute_one(
        conn,
        "INSERT INTO podcasts(file_id, title, turns, cast_json, audio_file_id)
         VALUES (?1, ?2, ?3, ?4, NULL)
         ON CONFLICT(file_id) DO UPDATE SET title = ?2, turns = ?3, cast_json = ?4",
        params![file_id, title, turns, cast],
    )
}

/// Replace the cast — the user re-casting a host, or changing a voice.
///
/// Deliberately does NOT clear `audio_file_id`. The previously rendered episode
/// is still a real file the user may be listening to; silently unlinking it
/// would make it look deleted. The UI compares the two and offers a re-record,
/// which is a choice rather than a surprise.
pub fn set_podcast_cast(
    conn: &Connection,
    file_id: &str,
    cast: &[PodcastHost],
) -> Result<(), String> {
    let cast = serde_json::to_string(cast).map_err(|e| e.to_string())?;
    execute_one(
        conn,
        "UPDATE podcasts SET cast_json = ?2 WHERE file_id = ?1",
        params![file_id, cast],
    )
}

/// Point a script at the episode that was rendered from it.
pub fn set_podcast_audio(
    conn: &Connection,
    file_id: &str,
    audio_file_id: &str,
) -> Result<(), String> {
    execute_one(
        conn,
        "UPDATE podcasts SET audio_file_id = ?2 WHERE file_id = ?1",
        params![file_id, audio_file_id],
    )
}

/// Build the opening cast from a script's turns.
///
/// TWO THINGS THIS FIXES, both of which break voices rather than looking wrong:
///
///   * SPEAKER DRIFT. Names are matched case- and whitespace-insensitively, so
///     "ada", "Ada" and "Ada " are one host with one voice, not three.
///   * VOICE COLLISION. Two hosts reading in the same voice is not a
///     two-voice podcast, so distinct voices are handed out in order and a
///     voice is never reused while an unused one remains.
///
/// `available` is the catalog to draw from, best-first — the caller decides
/// what "best" means (multilingual for a non-English script, alternating
/// genders when the catalog reports them). An empty catalog yields hosts with
/// empty voice ids, which means "the product default" everywhere else, so an
/// offline first-open still produces a usable cast the user can edit.
pub fn cast_from_turns(turns: &[PodcastTurn], available: &[String]) -> Vec<PodcastHost> {
    let mut hosts: Vec<PodcastHost> = Vec::new();
    for t in turns {
        let name = t.speaker.trim();
        if name.is_empty() {
            continue;
        }
        if hosts.iter().any(|h| h.name.eq_ignore_ascii_case(name)) {
            continue;
        }
        let voice = available.get(hosts.len()).cloned().unwrap_or_default();
        hosts.push(PodcastHost {
            name: name.to_string(),
            voice,
            rate: String::new(),
            pitch: String::new(),
        });
    }
    hosts
}

/// Fold every turn's speaker onto the cast's spelling of that name.
///
/// Called on the way IN, so the stored turns and the stored cast agree
/// letter-for-letter. Everything downstream — the voice lookup, the rendered
/// transcript's "Name:" labels, the panel's per-host preview — joins on this
/// string, and a mismatch is not a cosmetic difference: it is a line that gets
/// the default voice while the panel shows it assigned to someone else.
pub fn normalize_turn_speakers(turns: &mut [PodcastTurn], cast: &[PodcastHost]) {
    for t in turns.iter_mut() {
        if let Some(host) = cast
            .iter()
            .find(|h| h.name.eq_ignore_ascii_case(t.speaker.trim()))
        {
            t.speaker = host.name.clone();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn turn(speaker: &str, line: &str) -> PodcastTurn {
        PodcastTurn {
            speaker: speaker.into(),
            line: line.into(),
        }
    }

    #[test]
    fn a_cast_has_one_host_per_speaker_however_it_was_spelled() {
        let turns = vec![
            turn("Ada", "Welcome in."),
            turn("Bo", "Glad to be here."),
            turn("ada ", "Let's start."),
        ];
        let cast = cast_from_turns(&turns, &["v1".into(), "v2".into(), "v3".into()]);
        assert_eq!(cast.len(), 2, "case and spacing do not make a third host");
        assert_eq!(cast[0].name, "Ada");
        assert_eq!(cast[1].name, "Bo");
    }

    #[test]
    fn no_two_hosts_share_a_voice() {
        // Two hosts in one voice is not a two-voice podcast — it is one
        // narrator reading a dialogue, which is the thing this feature exists
        // to stop being.
        let turns = vec![turn("Ada", "a"), turn("Bo", "b"), turn("Cy", "c")];
        let cast = cast_from_turns(&turns, &["v1".into(), "v2".into(), "v3".into()]);
        let voices: Vec<&str> = cast.iter().map(|h| h.voice.as_str()).collect();
        assert_eq!(voices, vec!["v1", "v2", "v3"]);
    }

    #[test]
    fn an_empty_catalog_still_produces_an_editable_cast() {
        // Offline on first open: the hosts exist with the product-default
        // voice, so the panel has rows the user can change rather than nothing.
        let turns = vec![turn("Ada", "a"), turn("Bo", "b")];
        let cast = cast_from_turns(&turns, &[]);
        assert_eq!(cast.len(), 2);
        assert!(cast.iter().all(|h| h.voice.is_empty()));
    }

    #[test]
    fn turn_speakers_are_folded_onto_the_casts_spelling() {
        // The join key. Left unfolded, "ada" gets the default voice while the
        // panel shows Ada assigned to a chosen one.
        let mut turns = vec![turn("ada", "a"), turn("BO", "b"), turn("Ada", "c")];
        let cast = cast_from_turns(&turns, &["v1".into(), "v2".into()]);
        normalize_turn_speakers(&mut turns, &cast);
        assert_eq!(
            turns.iter().map(|t| t.speaker.as_str()).collect::<Vec<_>>(),
            vec!["ada", "BO", "ada"],
            "every line uses the cast's own spelling of its speaker",
        );
    }

    #[test]
    fn a_script_round_trips_through_the_room() {
        let conn = open_in_memory_schema();
        let file = insert_file(&conn, "ep.html", "text/html", b"<p>x</p>", None, "generated")
            .unwrap();
        let turns = vec![turn("Ada", "Welcome."), turn("Bo", "Hello.")];
        let cast = cast_from_turns(&turns, &["v1".into(), "v2".into()]);
        save_podcast(&conn, &file.id, "Episode 1", &turns, &cast).unwrap();

        let got = get_podcast(&conn, &file.id).unwrap();
        assert_eq!(got.title, "Episode 1");
        assert_eq!(got.turns, turns);
        assert_eq!(got.cast, cast);
        assert!(got.audio_file_id.is_none(), "nothing recorded yet");

        // Re-casting keeps the rendered audio linked — the old episode is a
        // real file the user may still be playing.
        let audio =
            insert_file(&conn, "ep.m4a", "audio/mp4", b"\0", None, "generated").unwrap();
        set_podcast_audio(&conn, &file.id, &audio.id).unwrap();
        let mut recast = cast.clone();
        recast[0].voice = "other".into();
        set_podcast_cast(&conn, &file.id, &recast).unwrap();
        let got = get_podcast(&conn, &file.id).unwrap();
        assert_eq!(got.cast[0].voice, "other");
        assert_eq!(got.audio_file_id.as_deref(), Some(audio.id.as_str()));
    }

    #[test]
    fn a_host_does_not_read_their_own_name_aloud() {
        // The reported bug: the model states the name as the field AND again
        // inside the text, so the voice service is handed "Alex: welcome in".
        assert_eq!(strip_speaker_label("Alex: welcome in", "Alex"), "welcome in");
        assert_eq!(strip_speaker_label("ALEX:  welcome in", "Alex"), "welcome in");
        assert_eq!(strip_speaker_label("**Jordan:** sure", "Jordan"), "sure");
        assert_eq!(strip_speaker_label("[Jordan]: sure", "Jordan"), "sure");
        // `Ada Lovelace` labelled as `Ada:` — the first word still matches.
        assert_eq!(strip_speaker_label("Ada: hello", "Ada Lovelace"), "hello");
        // Doubled, which a re-generated script really does produce.
        assert_eq!(strip_speaker_label("Alex: Alex: hi", "Alex"), "hi");
        // Already clean: unchanged, so the strip is safe to run on every read.
        assert_eq!(strip_speaker_label("welcome in", "Alex"), "welcome in");
    }

    #[test]
    fn an_ordinary_sentence_containing_a_colon_survives() {
        // THE RISK THIS FIX INTRODUCES. "cut at the first colon" would eat the
        // start of real sentences — a worse bug than the one being fixed, and a
        // silent one. Only a prefix that IS this speaker's name may go.
        assert_eq!(
            strip_speaker_label("Here's the thing: it works", "Alex"),
            "Here's the thing: it works",
        );
        // Another host's name is left alone — that is a mislabelled turn, and
        // guessing at it would rewrite what the script says.
        assert_eq!(strip_speaker_label("Jordan: no", "Alex"), "Jordan: no");
        // A colon far into prose is prose, not a label.
        let long = "One thing I keep coming back to about all of this is: it works";
        assert_eq!(strip_speaker_label(long, "Alex"), long);
        // A line that is ONLY a label strips to empty; callers decide.
        assert_eq!(strip_speaker_label("Alex:", "Alex"), "");
    }

    #[test]
    fn a_script_stored_before_the_strip_existed_is_fixed_when_it_is_read() {
        // No migration: the rows already in people's rooms are cleaned where
        // they are read, so re-recording an old episode stops speaking names.
        let conn = open_in_memory_schema();
        let file =
            insert_file(&conn, "ep.html", "text/html", b"<p>x</p>", None, "generated").unwrap();
        let dirty = vec![
            turn("Alex", "Alex: welcome in"),
            turn("Jordan", "Jordan: glad to be here"),
        ];
        let cast = cast_from_turns(&dirty, &["v1".into(), "v2".into()]);
        save_podcast(&conn, &file.id, "Episode 1", &dirty, &cast).unwrap();

        let got = get_podcast(&conn, &file.id).unwrap();
        assert_eq!(
            got.turns.iter().map(|t| t.line.as_str()).collect::<Vec<_>>(),
            vec!["welcome in", "glad to be here"],
        );
        // The speakers themselves are untouched — they are the join key.
        assert_eq!(got.turns[0].speaker, "Alex");
    }

    #[test]
    fn a_file_that_is_not_a_podcast_reads_as_none_not_an_error() {
        let conn = open_in_memory_schema();
        assert!(get_podcast(&conn, "no-such-file").is_none());
    }

    #[test]
    fn trashing_the_script_for_good_takes_its_podcast_row() {
        // ON DELETE CASCADE — a destroyed script must not leave a row pointing
        // at a file id that no longer resolves.
        let conn = open_in_memory_schema();
        let file =
            insert_file(&conn, "ep.html", "text/html", b"<p>x</p>", None, "generated").unwrap();
        save_podcast(&conn, &file.id, "t", &[turn("A", "x")], &[]).unwrap();
        delete_file(&conn, &file.id).unwrap();
        assert!(get_podcast(&conn, &file.id).is_none());
    }
}

File: apps/desktop/src/main/db-host/podcasts.ts
/**
 * A podcast script as data — port of `src-tauri/src/db/podcasts.rs` (438
 * lines, read in full including its `#[cfg(test)] mod tests`).
 *
 * The `podcasts` table already exists in `schema.sql`/`migrate.ts` (an
 * earlier batch created it, ready for whoever ported this module); nothing
 * here touches the schema.
 *
 * THE BUG THIS MODULE EXISTS FOR (see {@link stripSpeakerLabel}'s own doc):
 * a model states a turn's speaker twice — once as the `speaker` field, once
 * again inside the `line` text ("Ada: welcome in") — because that is what
 * dialogue looks like in its training data. Stripped at the read/write
 * boundary here, once, so a host never reads their own name aloud, the page
 * never prints the name twice, and the transcript never doubles it either.
 *
 * ESTABLISHED CONVENTIONS reused from `util.ts`: `queryOpt`/`queryRows`/
 * `executeOne`, positional `?` placeholders (a Rust `?1`/`?2` reused across a
 * statement becomes the same JS value repeated in the params array — see
 * {@link savePodcast}), and `.raw()` row access by position matching Rust's
 * `r.get(i)`.
 *
 * `PodcastHost`/`PodcastTurn`/`Podcast` are NOT redeclared — they already
 * exist, field-for-field, in `../../shared/apiTypes.ts` (and
 * `../../shared/ipc-contract.ts` already pins `get_podcast`/`set_podcast_cast`'s
 * exact args/result shapes), the same convention `peaksTools.ts`'s `AudioPeaks`
 * import documents: importing rather than inventing a second copy that could
 * drift from what the (not-yet-wired) preload/viewer side already expects.
 *
 * `eqIgnoreAsciiCase` is a small local predicate, not imported from anywhere
 * (nothing in this port exports one yet) — this port's established convention
 * for a small shared predicate with no canonical home (see `organizeTools.ts`'s
 * local `extensionOf`).
 */

import type Database from "better-sqlite3-multiple-ciphers";
import { executeOne, queryOpt, type Row } from "./util.js";
import type { Podcast, PodcastHost, PodcastTurn } from "../../shared/apiTypes.js";

// Re-exported (see this module's own doc, above) so every import site in this
// directory can pull the types from here without a second import line.
export type { Podcast, PodcastHost, PodcastTurn };

// =====================================================================
// str::eq_ignore_ascii_case — ASCII-only folding
// =====================================================================

/** `str::eq_ignore_ascii_case` — ASCII-only folding. Deliberately NOT
 * `toLowerCase()`: JS folds the full Unicode table, so `K` (U+212A) would
 * equal `k`, which is not what Rust's byte-for-byte ASCII fold reports. */
function asciiLowerCode(code: number): number {
  return code >= 65 && code <= 90 ? code + 32 : code;
}

export function eqIgnoreAsciiCase(a: string, b: string): boolean {
  if (a.length !== b.length) {
    return false;
  }
  for (let i = 0; i < a.length; i++) {
    if (asciiLowerCode(a.charCodeAt(i)) !== asciiLowerCode(b.charCodeAt(i))) {
      return false;
    }
  }
  return true;
}

// =====================================================================
// strip_speaker_label — the bug fix
// =====================================================================

/** Decoration a model wraps a speaker label in: `**Ada:**`, `[Ada]:`,
 * `> Ada:`. Ported verbatim from `DECOR`. */
const DECOR = new Set(["*", "_", "[", "]", "#", ">", '"', "'", "“", "”", " ", "\t"]);

/** A speaker label is short. Past this (in UTF-8 BYTES, matching Rust's
 * `char_indices` bound), a colon belongs to prose. */
const MAX_LABEL_BYTES = 48;

function trimStartDecor(s: string): string {
  let i = 0;
  while (i < s.length && DECOR.has(s[i] as string)) {
    i++;
  }
  return s.slice(i);
}

function trimDecor(s: string): string {
  let start = 0;
  let end = s.length;
  while (start < end && DECOR.has(s[start] as string)) {
    start++;
  }
  while (end > start && DECOR.has(s[end - 1] as string)) {
    end--;
  }
  return s.slice(start, end);
}

/** `char::is_whitespace()`-equivalent leading trim — Unicode `White_Space`,
 * not just ASCII `\s`. Used only for `line.trim_start()`. */
function trimStartUnicode(s: string): string {
  return s.replace(/^\p{White_Space}+/u, "");
}

/**
 * Byte length of a leading `Name:` run, or `null` if the line doesn't open
 * with this speaker's own name. Ported verbatim from `label_len`, tracking
 * BOTH the UTF-8 byte offset (Rust's `take_while` bound) and the UTF-16
 * string index (what `text.slice()` needs) as it scans, so the returned
 * index is a JS string index used consistently by {@link stripSpeakerLabel}.
 */
function labelColon(text: string): number | null {
  let byteIdx = 0;
  let strIdx = 0;
  for (const ch of text) {
    if (byteIdx > MAX_LABEL_BYTES) {
      break;
    }
    if (ch === ":") {
      return strIdx;
    }
    byteIdx += Buffer.byteLength(ch, "utf8");
    strIdx += ch.length;
  }
  return null;
}

function labelMatchesSpeaker(prefix: string, name: string): boolean {
  if (prefix === "") {
    return false;
  }
  const firstWord = name.split(/\p{White_Space}+/u).find((word) => word !== "") ?? name;
  return eqIgnoreAsciiCase(prefix, name) || eqIgnoreAsciiCase(prefix, firstWord);
}

function labelLen(text: string, speaker: string): number | null {
  const name = speaker.trim();
  if (name === "") {
    return null;
  }
  const colonAt = labelColon(text);
  if (colonAt === null) {
    return null;
  }
  const prefix = trimDecor(text.slice(0, colonAt));
  return labelMatchesSpeaker(prefix, name) ? colonAt + 1 : null;
}

/**
 * Drop a leading `Speaker:` label from a line — see this module's own doc
 * for the bug this exists to fix. Only ever strips a prefix that matches
 * THIS turn's own speaker (case-insensitively, ASCII-only), or that
 * speaker's first word, so an ordinary sentence containing a colon survives
 * untouched. Ported verbatim from `strip_speaker_label`.
 *
 * A line that is ONLY a label strips to `""`; callers decide what that
 * means.
 */
export function stripSpeakerLabel(line: string, speaker: string): string {
  let text = trimStartUnicode(line);
  // Bounded, not unbounded: `Ada: Ada: hi` is real (a re-generated script fed
  // its own previous output), but an unbounded loop on adversarial input is
  // not worth it.
  for (let i = 0; i < 2; i++) {
    const n = labelLen(text, speaker);
    if (n === null) {
      break;
    }
    text = trimStartDecor(text.slice(n));
  }
  return text;
}

/** Strip every turn's own speaker label, in place. A line that is nothing
 * BUT a label keeps its original text. Ported verbatim from
 * `strip_speaker_labels`. */
export function stripSpeakerLabels(turns: PodcastTurn[]): void {
  for (const t of turns) {
    const cleaned = stripSpeakerLabel(t.line, t.speaker);
    if (cleaned !== "") {
      t.line = cleaned;
    }
  }
}

// =====================================================================
// the table
// =====================================================================

/** Parse a `turns` JSON column the way `serde_json::from_str::<Vec<
 * PodcastTurn>>(..).unwrap_or_default()` does: not just "is it an array",
 * but "does every item have the required shape" — one malformed item fails
 * the WHOLE deserialization in Rust (a derived `Deserialize` on `Vec<T>`),
 * so this reports the whole column as empty rather than dropping just that
 * item. A row whose JSON will not parse this way is treated as an EMPTY
 * script rather than as a failed read (`podcast_row`'s own comment: the page
 * is still in the library and still opens; what is lost is the ability to
 * record it). */
function parseJsonArray(raw: string): unknown[] | null {
  try {
    const parsed: unknown = JSON.parse(raw);
    return Array.isArray(parsed) ? parsed : null;
  } catch {
    return null;
  }
}

function turnFrom(item: unknown): PodcastTurn | null {
  const value = objectRecord(item);
  if (value === null) {
    return null;
  }
  const speaker = requiredString(value, "speaker");
  const line = requiredString(value, "line");
  return speaker === null || line === null ? null : { speaker, line };
}

function parseTurnsColumn(raw: string): PodcastTurn[] {
  const parsed = parseJsonArray(raw);
  if (parsed === null) {
    return [];
  }
  const out: PodcastTurn[] = [];
  for (const item of parsed) {
    const turn = turnFrom(item);
    if (turn === null) {
      return [];
    }
    out.push(turn);
  }
  return out;
}

/** Same shape-validating parse as {@link parseTurnsColumn}, for the
 * `cast_json` column (`Vec<PodcastHost>` — `name` required, the rest
 * `#[serde(default)]`). */
function objectRecord(item: unknown): Record<string, unknown> | null {
  return typeof item === "object" && item !== null ? (item as Record<string, unknown>) : null;
}

function requiredString(value: Record<string, unknown>, name: string): string | null {
  const field = value[name];
  return typeof field === "string" ? field : null;
}

function optionalHostField(value: Record<string, unknown>, name: string): string {
  const field = value[name];
  return typeof field === "string" ? field : "";
}

function castHostFrom(item: unknown): PodcastHost | null {
  const value = objectRecord(item);
  if (value === null) {
    return null;
  }
  const name = requiredString(value, "name");
  if (name === null) {
    return null;
  }
  return {
    name,
    voice: optionalHostField(value, "voice"),
    rate: optionalHostField(value, "rate"),
    pitch: optionalHostField(value, "pitch"),
  };
}

function parseCastColumn(raw: string): PodcastHost[] {
  const parsed = parseJsonArray(raw);
  if (parsed === null) {
    return [];
  }
  const out: PodcastHost[] = [];
  for (const item of parsed) {
    const host = castHostFrom(item);
    if (host === null) {
      return [];
    }
    out.push(host);
  }
  return out;
}

function podcastRow(r: Row): Podcast {
  // On the way OUT as well as in, so scripts written before the label strip
  // existed are fixed where they are read instead of needing a migration.
  const turns = parseTurnsColumn(r[2] as string);
  stripSpeakerLabels(turns);
  return {
    fileId: r[0] as string,
    title: r[1] as string,
    turns,
    cast: parseCastColumn(r[3] as string),
    audioFileId: r[4] as string | null,
    createdAt: r[5] as string,
  };
}

/** The podcast belonging to a script page, if it has one. `null` covers two
 * real cases the caller must not conflate with an error: a file that is not
 * a podcast at all, and a podcast PAGE generated before this table existed
 * (or by the old HTML-authoring path), which has no turns to record. Ported
 * verbatim from `get_podcast`. */
export function getPodcast(db: Database.Database, fileId: string): Podcast | null {
  return queryOpt(
    db,
    "SELECT file_id, title, turns, cast_json, audio_file_id, created_at FROM podcasts WHERE file_id = ?",
    [fileId],
    podcastRow
  );
}

/** Store (or replace) a script's structure. Called by the studio's
 * `after_save` hook with the model's own JSON already parsed. Ported
 * verbatim from `save_podcast`. */
export function savePodcast(
  db: Database.Database,
  fileId: string,
  title: string,
  turns: readonly PodcastTurn[],
  cast: readonly PodcastHost[]
): void {
  const turnsJson = JSON.stringify(turns);
  const castJson = JSON.stringify(cast);
  executeOne(
    db,
    `INSERT INTO podcasts(file_id, title, turns, cast_json, audio_file_id)
     VALUES (?, ?, ?, ?, NULL)
     ON CONFLICT(file_id) DO UPDATE SET title = ?, turns = ?, cast_json = ?`,
    [fileId, title, turnsJson, castJson, title, turnsJson, castJson]
  );
}

/** Replace the cast — the user re-casting a host, or changing a voice.
 * Deliberately does NOT clear `audio_file_id` (see the Rust source's own
 * comment: the previously rendered episode is still a real file the user may
 * be listening to). Ported verbatim from `set_podcast_cast`. */
export function setPodcastCast(db: Database.Database, fileId: string, cast: readonly PodcastHost[]): void {
  const castJson = JSON.stringify(cast);
  executeOne(db, "UPDATE podcasts SET cast_json = ? WHERE file_id = ?", [castJson, fileId]);
}

/** Point a script at the episode that was rendered from it. Ported verbatim
 * from `set_podcast_audio`. */
export function setPodcastAudio(db: Database.Database, fileId: string, audioFileId: string): void {
  executeOne(db, "UPDATE podcasts SET audio_file_id = ? WHERE file_id = ?", [audioFileId, fileId]);
}

/**
 * Build the opening cast from a script's turns: one host per DISTINCT
 * speaker (case/whitespace-insensitive), each handed the next unused voice
 * from `available` (best-first) so no two hosts share a voice. An empty
 * catalog yields hosts with empty voice ids ("the product default"). Ported
 * verbatim from `cast_from_turns`.
 */
export function castFromTurns(turns: readonly PodcastTurn[], available: readonly string[]): PodcastHost[] {
  const hosts: PodcastHost[] = [];
  for (const t of turns) {
    const name = t.speaker.trim();
    if (name === "") {
      continue;
    }
    if (hosts.some((h) => eqIgnoreAsciiCase(h.name, name))) {
      continue;
    }
    const voice = available[hosts.length] ?? "";
    hosts.push({ name, voice, rate: "", pitch: "" });
  }
  return hosts;
}

/** Fold every turn's speaker onto the cast's spelling of that name, in
 * place — the join key every voice lookup, transcript label and panel
 * preview depends on. Ported verbatim from `normalize_turn_speakers`. */
export function normalizeTurnSpeakers(turns: PodcastTurn[], cast: readonly PodcastHost[]): void {
  for (const t of turns) {
    const host = cast.find((h) => eqIgnoreAsciiCase(h.name, t.speaker.trim()));
    if (host !== undefined) {
      t.speaker = host.name;
    }
  }
}
